### ISSUE TITLE
Some imputeLearners generate errors

### ISSUE THREAD

**Issue by ecrutchfield** [mlr-org/mlr#663]
Good afternoon,
I have been trying to impute data using the impute() method, but have seen inconsistent results across different classes. Maybe I misunderstand the "missings" property on the listLearners() function, but I thought that property would tell me which packages I could use for imputing missing data. Of the 8 classes that I have which can deal with missing data, only 3 will work without error. Here are the results:

Class | Package | Result
----- | ------- | ------
regr.bartMachine | bartMachine | Error in build_bart_machine(X, y, Xy, num_trees, num_burn_in, num_iterations_after_burn_in,  :   You cannot have any missing data in your response vector. 
regr.blackboost | mboost,party | Works
regr.cforest | party | Error in model@dpp(...) : missing values in response variable not allowed 
regr.ctree         |  party |  Error in model@dpp(...) : missing values in response variable not allowed 
regr.cubist        |  Cubist |  Error in strsplit(tmp, "\"")[[1]] : subscript out of bounds 
regr.gbm          |   gbm | Error in checkMissing(x, y) :   Missing values are not allowed in the response 
regr.randomForestSRC | randomForestSRC | Works
regr.rpart         |  rpart | Works

Here's the code I ran:
```

library(testthat)
library(mlr)
data(airquality)
airq = airquality
ind = sample(nrow(airq), 10)
airq$Wind[ind] = NA
airq$Wind = cut(airq$Wind, c(0,8,16,24))
imp = impute(airq, classes = list(integer = imputeMean(), factor = imputeMode()),
             dummy.classes = "integer")
expect_equal(length(which(is.na(imp$data$Solar.R))),0)
expect_equal(length(which(is.na(imp$data$Ozone))),0)
expect_equal(length(which(is.na(imp$data$Wind))),0)

# airq = subset(airq, select = 1:4)
airq.train = airq[1:100,]
airq.test = airq[-c(1:100),]

listLearners("regr", properties = "missings")[,c("class","package")]

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.bartMachine"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.blackboost"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.cforest"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.ctree"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.cubist"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.gbm"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.cubist"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.randomForestSRC"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)

imp = NULL
imp = impute(airq.train, target="Ozone", cols = list(Solar.R = imputeLearner(makeLearner("regr.rpart"))))
expect_equal(length(which(is.na(imp$data$Solar.R))),0)
```

Cheers,
Eric

**Comment by larskotthoff** [comment]
The `missings` properties refers only to features, not to response values. That is, all learners that have the `missings` property can deal with incomplete feature matrices, but still need labels for every observation. Similarly, the imputation functions only impute features.

**Comment by schiffner** [comment]
Hi Eric,
thanks for the report.
To clarify things a bit:
* The `missings` property simply says that a learner can deal with missing values in the features (i.e. predictor/explanatory/independent variables) on its own, e.g. through imputation and/or surrogate variables.
* You can in principle use any learner (suitable for the type of variable you want to impute) for imputation,   not just those with property `missings`.
* One important restriction though: If you want to impute one feature based on several others and at least one of those has missing values, you either have to impute these features first by some other function like `imputeMean` or you have to use a learner with property `missings` for imputation.
* The problem you observed with `imputeLearner` is due to the fact that it currently does not remove the missing values from the target variable (i.e. the variable you want to impute) before training the learner. I mentioned it in #26 and it's basically my fault that it's not fixed yet. :(
Will try to do this soon.

Cheers,
Julia

**Comment by ecrutchfield** [comment]
Ah, thanks for the heads up on #26 @schiffner! I did not see that when I was looking for existing issues.

I have the case you mention in your third bullet (imputing features using other features with missing data), so it seems I am appropriately restricting my learners to those that support missing data. 

Anyway, I think I have enough options with the 3 working imputeLearner classes that I can move forward without any problems.

Thanks for the quick response!

Cheers,
Eric

**Comment by berndbischl** [comment]
Hi,

how shall we clean this issue up?

1) Do we have a little table somewhere in the docs / learner API in the tutorial where all of properties are explained? If not we should add it.

2) In Erics code, is any error now not a duplicate of #26?

**Comment by schiffner** [comment]
> 1) Do we have a little table somewhere in the docs / learner API in the tutorial where all of properties are explained? If not we should add it.

In the docs there is a table in `?RLearner`, which I suppose  is not the most prominent place where people look. I guess we should doc the `Learner` object in `?makeLearner` and also include a table with properties there (preferably a reusable template).

In the tutorial there are some explanations on the page with the learner tables (http://mlr-org.github.io/mlr-tutorial/devel/html/integrated_learners/index.html), which can be improved a bit. Also the new create-learner page will have some more explanations (I'm not done yet).
I thought about putting a table on the learners page, but maybe it's not necessary if it's in the docs.

So bottomline: We doc it in `makeLearner` and I will look where I can make the tutorial better.

> 2) In Erics code, is any error now not a duplicate of #26?

No.

**Comment by berndbischl** [comment]
@schiffner 

1) I have updated the properties docs  (weights was missing) and moved them to ?LearnerProperties.
I linked to this page from makeLearner and RLearner. 

2) Would you please:
- Reread what I did and check?
- Make sure that the properties are visible in the "Extend Learner" section of the tutorial. You could link to LearnerProperties, but I guess here it is better to be redundant and explain this.

Then we can close here.

My commits are here:
https://github.com/mlr-org/mlr/commit/1d4fe9598c8a41be0cf87be4023faf43765e912d
https://github.com/mlr-org/mlr/commit/ed71e529165a447f26860ce2569d4ca022b79527

**Comment by schiffner** [comment]
@berndbischl Thanks for doing all this.
Sorry about `weights`, but good that you noticed. :) I'm reviewing now.

**Comment by schiffner** [comment]
https://github.com/mlr-org/mlr/blob/ed71e529165a447f26860ce2569d4ca022b79527/R/Learner_properties.R#L15: prob is also applicable to multilabel and surv

**Comment by schiffner** [comment]
That was the only thing I noticed. Thanks again.

**Comment by berndbischl** [comment]
Ok, so I fix the "prob" thing, the tutorial is OK in your opionion, and we close here?

**Comment by schiffner** [comment]
> ..., the tutorial is OK in your opionion, and we close here?

I would say it's not totally OK at the moment, but I'm working on it and we can close here anyway.

**Comment by kapelner** [comment]
I am the maintainer of bartMachine. You need to pass in use_missing_data = TRUE in order for bartMachine to handle missing data natively. By default this is turned off. Unsure how this is handled inside mlr. It is possible mlr wants to turn it on by default. 

I'm releasing a new version that will fix a lot of previous issues with bartMachine soon.

**Comment by kapelner** [comment]
Just to clarify: bartMachine does not impute... it handles missingness natively. If you want to use it to impute, you can run a algorithm similar to missForest using bartMachine. We have not done so but we believe it to be a nice contribution.

**Comment by schiffner** [comment]
> You need to pass in use_missing_data = TRUE in order for bartMachine to handle missing data natively. By default this is turned off. Unsure how this is handled inside mlr. It is possible mlr wants to turn it on by default.

Thx. I checked the learner definition. use_missing_data is set to TRUE by default.

**Comment by schiffner** [comment]
> The problem you observed with `imputeLearner` is due to the fact that it currently does not remove the missing values from the target variable (i.e. the variable you want to impute) before training the learner. I mentioned it in #26 and it's basically my fault that it's not fixed yet. :( Will try to do this soon.

This has been resolved in #1036.
